### PR TITLE
[TOSA] Add aten.Index.Tensor support

### PR DESCRIPTION
**Pull request by AmosLewis**
Find this error in https://github.com/nod-ai/SHARK/issues/494
**torch.ops.aten.index(input, (index, ))** 
```
  %530 = torch_c.from_builtin_tensor %24 : tensor<1xi64> -> !torch.vtensor<[1],si64>
  %531 = "tosa.slice"(%529) {size = [1, 1, 2], start = [0, 127, 0]} : (tensor<1x128x2xf32>) -> tensor<1x1x2xf32>
  %532 = "tosa.reshape"(%531) {new_shape = [1, 2]} : (tensor<1x1x2xf32>) -> tensor<1x2xf32>
  %533 = torch_c.from_builtin_tensor %532 : tensor<1x2xf32> -> !torch.vtensor<[1,2],f32>
  %534 = torch.prim.ListConstruct %530 : (!torch.vtensor<[1],si64>) -> !torch.list<vtensor>
  %535 = torch.aten.index.Tensor %533, %534 : !torch.vtensor<[1,2],f32>, !torch.list<vtensor> -> !torch.vtensor<[1,2],f32>
  %536 = torch_c.to_builtin_tensor %535 : !torch.vtensor<[1,2],f32> -> tensor<1x2xf32>
```
```
func.func @torch.aten.index.Tensor(%arg0: !torch.vtensor<[1],si64>, %arg1: !torch.vtensor<[1,2],f32>) -> !torch.vtensor<[1,2],f32> {
  %0 = torch.prim.ListConstruct %arg0 : (!torch.vtensor<[1],si64>) -> !torch.list<vtensor>
  %1 = torch.aten.index.Tensor %arg1, %0 : !torch.vtensor<[1,2],f32>, !torch.list<vtensor> -> !torch.vtensor<[1,2],f32>
  return %1 : !torch.vtensor<[1,2],f32>
}
```

We could use the tf gather_nd(which is already implemented in tosa) to get the torch.index. The trick is to expand/unsqueeze the last dim of the index. 
python code for algorithm explanation:
```
import torch

t = torch.tensor([ [1, 2, 3, 4, 5], [6,7,8,9,10],[11,12,13,14,15],[16,17,18,19,20]]) # 4*5
i = torch.tensor([[1,2,3], [3,2,1]]) # 2*3

o = t[i] 
  = torch.tensor([ [[ 6,  7,  8,  9, 10], [11, 12, 13, 14, 15], [16, 17, 18, 19, 20]],  
[[16, 17, 18, 19, 20], [11, 12, 13, 14, 15], [ 6,  7,  8,  9, 10]]]) #  2*3*5
  

########## same function by tensorflow
import tensorflow as tf

t = tf.constant([[1, 2, 3, 4, 5], [6,7,8,9,10],[11,12,13,14,15],[16,17,18,19,20]]) # 4*5

i =  tf.constant([[1,2,3], [3,2,1]]) # 2*3

i_expand = tf.expand_dims(i,axis=2)
<tf.Tensor: shape=(2, 3, 1), dtype=int32, numpy=
array([[[1],[2],[3]],[[3],[2],[1]]], dtype=int32)>

io=tf.gather_nd(t,tf.expand_dims(i,axis=2))
<tf.Tensor: shape=(2, 3, 5), dtype=int32, numpy=
array([[[ 6,  7,  8,  9, 10],[11, 12, 13, 14, 15],[16, 17, 18, 19, 20]],
       [[16, 17, 18, 19, 20], [11, 12, 13, 14, 15],[ 6,  7,  8,  9, 10]]], dtype=int32)>
```

Final result should looks like this:
```
// -----
// CHECK-LABEL:   func.func @torch.aten.index.Tensor(
// CHECK-SAME:                                       %[[VAL_0:.*]]: !torch.vtensor<[1],si64>,
// CHECK-SAME:                                       %[[VAL_1:.*]]: !torch.vtensor<[1,2],f32>) -> !torch.vtensor<[1,2],f32> {
// CHECK:           %[[VAL_2:.*]] = torch_c.to_builtin_tensor %[[VAL_1]] : !torch.vtensor<[1,2],f32> -> tensor<1x2xf32>
// CHECK:           %[[VAL_3:.*]] = torch.prim.ListConstruct %[[VAL_0]] : (!torch.vtensor<[1],si64>) -> !torch.list<vtensor>
// CHECK:           %[[VAL_4:.*]] = torch_c.to_builtin_tensor %[[VAL_0]] : !torch.vtensor<[1],si64> -> tensor<1xi64>
// CHECK:           %[[VAL_5:.*]] = "tosa.cast"(%[[VAL_4]]) : (tensor<1xi64>) -> tensor<1xi32>
// CHECK:           %[[VAL_6:.*]] = "tosa.reshape"(%[[VAL_5]]) {new_shape = array<i64: 1, 1>} : (tensor<1xi32>) -> tensor<1x1xi32>
// CHECK:           %[[VAL_7:.*]] = "tosa.reshape"(%[[VAL_2]]) {new_shape = array<i64: 1, 1, 2>} : (tensor<1x2xf32>) -> tensor<1x1x2xf32>
// CHECK:           %[[VAL_8:.*]] = "tosa.reshape"(%[[VAL_6]]) {new_shape = array<i64: 1, 1>} : (tensor<1x1xi32>) -> tensor<1x1xi32>
// CHECK:           %[[VAL_9:.*]] = "tosa.const"() {value = dense<1> : tensor<1xi32>} : () -> tensor<1xi32>
// CHECK:           %[[VAL_10:.*]] = "tosa.mul"(%[[VAL_8]], %[[VAL_9]]) {shift = 0 : i32} : (tensor<1x1xi32>, tensor<1xi32>) -> tensor<1x1xi32>
// CHECK:           %[[VAL_11:.*]] = "tosa.reduce_sum"(%[[VAL_10]]) {axis = 1 : i64} : (tensor<1x1xi32>) -> tensor<1x1xi32>
// CHECK:           %[[VAL_12:.*]] = "tosa.reshape"(%[[VAL_11]]) {new_shape = array<i64: 1, 1>} : (tensor<1x1xi32>) -> tensor<1x1xi32>
// CHECK:           %[[VAL_13:.*]] = "tosa.gather"(%[[VAL_7]], %[[VAL_12]]) : (tensor<1x1x2xf32>, tensor<1x1xi32>) -> tensor<1x1x2xf32>
// CHECK:           %[[VAL_14:.*]] = "tosa.reshape"(%[[VAL_13]]) {new_shape = array<i64: 1, 2>} : (tensor<1x1x2xf32>) -> tensor<1x2xf32>
// CHECK:           %[[VAL_15:.*]] = torch_c.from_builtin_tensor %[[VAL_14]] : tensor<1x2xf32> -> !torch.vtensor<[1,2],f32>
// CHECK:           return %[[VAL_15]] : !torch.vtensor<[1,2],f32>
// CHECK:         }
func.func @torch.aten.index.Tensor(%arg0: !torch.vtensor<[1],si64>, %arg1: !torch.vtensor<[1,2],f32>) -> !torch.vtensor<[1,2],f32> {
  %0 = torch.prim.ListConstruct %arg0 : (!torch.vtensor<[1],si64>) -> !torch.list<vtensor>
  %1 = torch.aten.index.Tensor %arg1, %0 : !torch.vtensor<[1,2],f32>, !torch.list<vtensor> -> !torch.vtensor<[1,2],f32>
  return %1 : !torch.vtensor<[1,2],f32>
}
```

### CODE REVIEW

**Comment by AmosLewis** [comment]
e2e fail, working on fix it.

**Comment by AmosLewis** [comment]
Fix the e2e bug. 
The new implementation is explained here:
https://gist.github.com/AmosLewis/c90c1148a96291db93408b3fa39f9ae2

**Comment by AmosLewis** [comment]
Add a check to fix  IndexTensorMultiInputContiguousCenter e2e test segfault
```
// TODO figure out why the index is empty for IndexTensorMultiInputContiguousCenter e2e test
  if (!index.getImpl())
    return rewriter.notifyMatchFailure(
        op, "Only list ranked tensor types index are supported");
```

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp [hunk 1]
```diff
@@ -3232,6 +3233,74 @@ LogicalResult ConvertAtenOp<AtenGatherOp>::matchAndRewrite(
...
+  // Deal with torch.prim.ListConstruct of non const value to get the index
+  auto tensorList = op.getIndices();
+  SmallVector<Value> tensorsTorchType;
+  if (!getListConstructElements(tensorList, tensorsTorchType))
+    return op.emitError(
+        "unimplemented: the tensor list is not from list construct");
+  auto tensors = getTypeConvertedValues(rewriter, op->getLoc(),
+                                        getTypeConverter(), tensorsTorchType);
+
+  auto index = tensors[0];
```
Comment (eric-k256) [inline lib/Conversion/TorchToTosa/TorchToTosa.cpp:3259]: Is there a plan if the size of tensors is > 1? Return a failure with a message noting that it is currently unsupported? Looking at the Torch to linalg version, it appears that multiple index tensors adds quite a bit of complication. I'm not sure how often index lists are used.

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp [hunk 2]
```diff
@@ -3232,6 +3233,74 @@ LogicalResult ConvertAtenOp<AtenGatherOp>::matchAndRewrite(
...
+  // TODO figure out why the index is empty for
+  // IndexTensorMultiInputContiguousCenter e2e test
+  if (!index.getImpl())
+    return rewriter.notifyMatchFailure(
+        op, "Only list ranked tensor types index are supported");
+  auto indexType = index.getType().dyn_cast<RankedTensorType>();
```
[the remark this answers is not in the document]
Comment (AmosLewis) [reply lib/Conversion/TorchToTosa/TorchToTosa.cpp:3265]: I am pretty sure this will success. I have tried the dyn_cast<>, but it will segfault for [IndexTensorMultiInputContiguousCenter](https://gist.github.com/AmosLewis/144ff9fae33e1bf99cd3986326914cd4). Because to use dyn_cast, you will have to call index.getType(), which will get the *index.getImpl(). But sometimes it is void which will lead to segfault.

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp (hunk 1 above)
Comment (AmosLewis) [reply lib/Conversion/TorchToTosa/TorchToTosa.cpp:3259]: Add it to todo

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (AmosLewis) [reply lib/Conversion/TorchToTosa/TorchToTosa.cpp:3265]: example: torch.ops.aten.index(x, (None, index1, index2, None))

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp [hunk 3]
```diff
@@ -3232,6 +3233,74 @@ LogicalResult ConvertAtenOp<AtenGatherOp>::matchAndRewrite(
   return success();
 }
 
+template <>
+LogicalResult ConvertAtenOp<AtenIndexTensorOp>::matchAndRewrite(
+    AtenIndexTensorOp op, OpAdaptor adaptor,
+    ConversionPatternRewriter &rewriter) const {
+  // Python example of this algorithm:
+  // https://gist.github.com/AmosLewis/c90c1148a96291db93408b3fa39f9ae2
```
[the remark this answers is not in the document]
Comment (AmosLewis) [reply lib/Conversion/TorchToTosa/TorchToTosa.cpp]: Done

File: test/Conversion/TorchToTosa/basic.mlir [hunk 4]
```diff
@@ -970,6 +970,32 @@ func.func @torch.aten.gather(%arg0: !torch.vtensor<[1,4,3],f32>, %arg1: !torch.v
...
+// CHECK:           %[[VAL_9:.*]] = "tosa.const"() {value = dense<1> : tensor<1xi32>} : () -> tensor<1xi32>
+// CHECK:           %[[VAL_10:.*]] = "tosa.mul"(%[[VAL_8]], %[[VAL_9]]) {shift = 0 : i32} : (tensor<1x1xi32>, tensor<1xi32>) -> tensor<1x1xi32>
+// CHECK:           %[[VAL_11:.*]] = "tosa.reduce_sum"(%[[VAL_10]]) {axis = 1 : i64} : (tensor<1x1xi32>) -> tensor<1x1xi32>
+// CHECK:           %[[VAL_12:.*]] = "tosa.reshape"(%[[VAL_11]]) {new_shape = array<i64: 1, 1>} : (tensor<1x1xi32>) -> tensor<1x1xi32>
+// CHECK:           %[[VAL_13:.*]] = "tosa.gather"(%[[VAL_7]], %[[VAL_12]]) : (tensor<1x1x2xf32>, tensor<1x1xi32>) -> tensor<1x1x2xf32>
+// CHECK:           %[[VAL_14:.*]] = "tosa.reshape"(%[[VAL_13]]) {new_shape = array<i64: 1, 2>} : (tensor<1x1x2xf32>) -> tensor<1x2xf32>
+// CHECK:           %[[VAL_15:.*]] = torch_c.from_builtin_tensor %[[VAL_14]] : tensor<1x2xf32> -> !torch.vtensor<[1,2],f32>
+// CHECK:           return %[[VAL_15]] : !torch.vtensor<[1,2],f32>
+// CHECK:         }
+func.func @torch.aten.index.Tensor(%arg0: !torch.vtensor<[1],si64>, %arg1: !torch.vtensor<[1,2],f32>) -> !torch.vtensor<[1,2],f32> {
```
[the remark this answers is not in the document]
Comment (AmosLewis) [reply test/Conversion/TorchToTosa/basic.mlir]: Done

**Comment by ramiro050** [comment]
Removing myself from reviewers. I will let @eric-k256 continue leading the review of this patch (happy to help if questions/issues arise)

**Review by eric-k256 — APPROVED** [review]
